### ISSUE TITLE
Catch-all Segment works as Optional Catch-all Segments inside the Parallel Route

### ISSUE THREAD

**Issue by aramikuto** [vercel/next.js#60613]
### Link to the code that reproduces this issue

https://github.com/aramikuto/catch-all-inside-parallel-route-problem

### To Reproduce

1. Create a Catch-all Segment inside a Parallel Route.
2. Render the parallel route without adding any additional parameters.

### Current vs. Expected behavior

Current behavior:
The content of the Catch-all Segment is not rendered because it should not match empty parameters.

Expected behavior:
The content of the Catch-all Segment is rendered, and the type of the catch-all parameter is `undefined`.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.1.0: Mon Oct  9 21:28:12 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T8103
Binaries:
  Node: 18.19.0
  npm: 10.2.3
  Yarn: 1.22.21
  pnpm: 8.12.1
Relevant Packages:
  next: 14.0.5-canary.55
  eslint-config-next: 14.0.5-canary.38
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router, Routing (next/router, next/navigation, next/link)

### Which stage(s) are affected? (Select all that apply)

next dev (local), next start (local)

<sub>[NEXT-2243](https://linear.app/vercel/issue/NEXT-2243/catch-all-segment-works-as-optional-catch-all-segments-inside-the)</sub>

**Comment by maltesa** [comment]
I can confirm this issue. And I think it was open (and closed) once before #49531. 

Here is a reproduction in a codesandbox with the latest nextjs canary: https://codesandbox.io/p/devbox/angry-browser-z3mtgs

This is what happens:
https://github.com/vercel/next.js/assets/4067078/26ce6745-19a4-4abc-9b6e-6f9efab42da0

**Expected:**
Catch-All Route fires for `/foo` too.

**Actually:**
Catch-All Route fires for anything except `/foo` and `/`.

I expected a different behavior from reading the documentation: https://nextjs.org/docs/app/building-your-application/routing/parallel-routes#dismissing-a-modal

**Comment by aramikuto** [comment]
@maltesa Thank you for looking into the problem! I feel like the issue you've linked is a little bit different. It describes a case where the parallel route does not match in a situation where it should, whereas in my case, the parallel route matches a segment it should not.

It seems that in Next.js 13.5.6, my problem was not present: the `/` route does not match `[...all]`.
